### PR TITLE
Close #62: strange error output on recursive function calls

### LINKED ISSUE #62: strange error output on recursive function calls
Recursive calls are not allowed in shading languages, but the compiler should report a meaningful error in that case.

I use the following example code with the online HLSL compiler ([tryhlsl.azurewebsites.net](http://tryhlsl.azurewebsites.net/)):
```c
float4 f(); // Forward declaration
float4 g() { return f(); } // Recursive call to 'f'
float4 f() { return g(); } // First call to 'g'

float4 VS() : SV_Position
{
    return f(); // First call to 'f'
}
```
The error output of the old HLSL compiler with target "vs_5_1" is:
```
D:\Windows\system32\unknown(4,8-10): error X3500: 'f': recursive functions not allowed in vs_5_1
```

But the error output of the new HLSL compiler with target "vs_6_0" is:
```
error: validation errors
Function ?f@@YA?AV?$vector@M$03@@XZ.flat with parameter is not permitted, it should be inlined
Not all elements of output SV_Position were written
Not all elements of SV_Position were written
```

I assume that this is not the intended output.

### PR DESCRIPTION
**Pull request by marcelolr**
This set of changes addresses the error and includes some work that falls out of that.

The check for recursive calls now happens in the front-end, before constant folding and dead code elimination, and so the condition is no longer sensitive to optimization settings. The error message is also nicer, as it notes the offending function with its unmangled name.

Because this requires walking all reachable functions, a prior implementation tried to use this to avoid the codegen expense. A better way to address this (and "pay back" for the cost of the checks) is to properly implement the check as to whether functions should be emitted. The linkage specification doesn't really apply, and so the default of exporting functions (and thus requiring them to be code gen'ed) isn't appropriate.

There is also some test reorganization to effectively test the prior IR-based check and better reuse some verification code.

### CODE REVIEW

File: tools/clang/lib/AST/ASTContext.cpp [hunk 1]
```diff
@@ -8461,7 +8461,10 @@ bool ASTContext::DeclMustBeEmitted(const Decl *D) {
     if (Linkage == GVA_Internal || Linkage == GVA_AvailableExternally ||
         Linkage == GVA_DiscardableODR)
       return false;
-    return true;
+    // HLSL Change Starts
+    // Don't just return true because of visibility, unless building a library (which is not currently implemented)
+    return FD->getName() == getLangOpts().HLSLEntryFunction || IsPatchConstantFunctionDecl(FD);
```
Comment (python3kgae) [inline tools/clang/lib/AST/ASTContext.cpp]: Could we only check IsPatchConstantFunctionDecl for HullShader profile here?

File: tools/clang/lib/Sema/SemaHLSL.cpp [hunk 2]
```diff
@@ -8568,6 +8703,105 @@ void hlsl::DiagnoseRegisterType(
...
+        "missing entry point definition");
+      Diags.Report(id);
+      return;
+    }
+  }
+
+  // Validate that there is no recursion; start with the entry function.
+  // NOTE: the information gathered here could be used to bypass code generation
+  // on functions that are unreachable (as an early form of dead code elimination).
+  if (pEntryPointDecl) {
```
Comment (python3kgae) [inline tools/clang/lib/Sema/SemaHLSL.cpp]: Do we need to report error if pEntryPointDecl is nullptr here?

File: tools/clang/lib/AST/ASTContext.cpp (hunk 1 above)
Comment (marcelolr) [inline tools/clang/lib/AST/ASTContext.cpp]: Not right now. ASTContext doesn't have visibility into the target profile. We'll have to add it when we add library support (and then at that point we'll be revisiting this bit of code in any case).

File: tools/clang/lib/Sema/SemaHLSL.cpp (hunk 2 above)
Comment (marcelolr) [inline tools/clang/lib/Sema/SemaHLSL.cpp]: I don't think so - there's a different error that will kick in if we looked for the entry point but couldn't find it and we wouldn't reach here. We would hit this codepath if the entry point isn't specified at all, in which case we're probably running in an IDE, and then we shouldn't be running this bit of analysis.

**Pull request merged**